drivers/openxr/openxr_device.cpp

```diff
@@ -0,0 +1,2386 @@
+		ENUM_TO_STRING_CASE(XR_TYPE_PASSTHROUGH_CREATE_INFO_FB)
+		ENUM_TO_STRING_CASE(XR_TYPE_PASSTHROUGH_LAYER_CREATE_INFO_FB)
+		ENUM_TO_STRING_CASE(XR_TYPE_COMPOSITION_LAYER_PASSTHROUGH_FB)
+		ENUM_TO_STRING_CASE(XR_TYPE_GEOMETRY_INSTANCE_CREATE_INFO_FB)
+		ENUM_TO_STRING_CASE(XR_TYPE_GEOMETRY_INSTANCE_TRANSFORM_FB)
+		ENUM_TO_STRING_CASE(XR_TYPE_PASSTHROUGH_STYLE_FB)
+		ENUM_TO_STRING_CASE(XR_TYPE_PASSTHROUGH_COLOR_MAP_MONO_TO_RGBA_FB)
+		ENUM_TO_STRING_CASE(XR_TYPE_PASSTHROUGH_COLOR_MAP_MONO_TO_MONO_FB)
+		ENUM_TO_STRING_CASE(XR_TYPE_EVENT_DATA_PASSTHROUGH_STATE_CHANGED_FB)
+		ENUM_TO_STRING_CASE(XR_TYPE_BINDING_MODIFICATIONS_KHR)
+		ENUM_TO_STRING_CASE(XR_TYPE_VIEW_LOCATE_FOVEATED_RENDERING_VARJO)
+		ENUM_TO_STRING_CASE(XR_TYPE_FOVEATED_VIEW_CONFIGURATION_VIEW_VARJO)
+		ENUM_TO_STRING_CASE(XR_TYPE_SYSTEM_FOVEATED_RENDERING_PROPERTIES_VARJO)
+		ENUM_TO_STRING_CASE(XR_TYPE_COMPOSITION_LAYER_DEPTH_TEST_VARJO)
+		ENUM_TO_STRING_CASE(XR_TYPE_SYSTEM_MARKER_TRACKING_PROPERTIES_VARJO)
+		ENUM_TO_STRING_CASE(XR_TYPE_EVENT_DATA_MARKER_TRACKING_UPDATE_VARJO)
+		ENUM_TO_STRING_CASE(XR_TYPE_MARKER_SPACE_CREATE_INFO_VARJO)
+		ENUM_TO_STRING_CASE(XR_TYPE_SPATIAL_ANCHOR_PERSISTENCE_INFO_MSFT)
+		ENUM_TO_STRING_CASE(XR_TYPE_SPATIAL_ANCHOR_FROM_PERSISTED_ANCHOR_CREATE_INFO_MSFT)
+		ENUM_TO_STRING_CASE(XR_TYPE_SWAPCHAIN_IMAGE_FOVEATION_VULKAN_FB)
+		ENUM_TO_STRING_CASE(XR_TYPE_SWAPCHAIN_STATE_ANDROID_SURFACE_DIMENSIONS_FB)
+		ENUM_TO_STRING_CASE(XR_TYPE_SWAPCHAIN_STATE_SAMPLER_OPENGL_ES_FB)
+		ENUM_TO_STRING_CASE(XR_TYPE_SWAPCHAIN_STATE_SAMPLER_VULKAN_FB)
+		ENUM_TO_STRING_CASE(XR_TYPE_COMPOSITION_LAYER_SPACE_WARP_INFO_FB)
+		ENUM_TO_STRING_CASE(XR_TYPE_SYSTEM_SPACE_WARP_PROPERTIES_FB)
+		ENUM_TO_STRING_CASE(XR_STRUCTURE_TYPE_MAX_ENUM)
+		default: {
+			return String("Structure type ") + String::num_int64(int64_t(p_structure_type));
+		} break;
+	}
+}
+
+String OpenXRDevice::get_session_state_name(XrSessionState p_session_state) const {
+	switch (p_session_state) {
+		ENUM_TO_STRING_CASE(XR_SESSION_STATE_UNKNOWN)
+		ENUM_TO_STRING_CASE(XR_SESSION_STATE_IDLE)
+		ENUM_TO_STRING_CASE(XR_SESSION_STATE_READY)
+		ENUM_TO_STRING_CASE(XR_SESSION_STATE_SYNCHRONIZED)
+		ENUM_TO_STRING_CASE(XR_SESSION_STATE_VISIBLE)
+		ENUM_TO_STRING_CASE(XR_SESSION_STATE_FOCUSED)
+		ENUM_TO_STRING_CASE(XR_SESSION_STATE_STOPPING)
+		ENUM_TO_STRING_CASE(XR_SESSION_STATE_LOSS_PENDING)
+		ENUM_TO_STRING_CASE(XR_SESSION_STATE_EXITING)
+		ENUM_TO_STRING_CASE(XR_SESSION_STATE_MAX_ENUM)
+		default: {
+			return String("Session state ") + String::num_int64(int64_t(p_session_state));
+		} break;
+	}
+}
+
+String OpenXRDevice::make_xr_version_string(XrVersion p_version) {
+	String version;
+
+	version += String::num_int64(XR_VERSION_MAJOR(p_version));
+	version += String(".");
+	version += String::num_int64(XR_VERSION_MINOR(p_version));
+	version += String(".");
+	version += String::num_int64(XR_VERSION_PATCH(p_version));
+
+	return version;
+}
```

Comment:
It may be useful to move the above functions into a util header file to keep this class smaller and concise.

Reply:
I've moved those not dependent on member variables into a new `openxr_util.h` header file. Does indeed help a bit.